src/convnets/resnets.jl

```diff
@@ -0,0 +1,543 @@
+# resnet.jl
+## It is recommended to check out the user's guide for more information
+## regarding the use of these functions.
+
+### ResNet blocks
+## These functions return a block to be used inside of a ResNet model.
+## The individual arguments are explained in the documentation of the functions.
+## Note that for these blocks to be used by the `_make_blocks` function, they must define
+## a dispatch `expansion(::typeof(fn))` that returns the expansion factor of the block 
+## (i.e. the multiplicative factor by which the number of channels in the input is increased).
+## The `_make_blocks` function will then call the `expansion` function to determine the
+## expansion factor of each block and use this to construct the stages of the model.
+
+"""
+    basicblock(inplanes, planes; stride = 1, downsample = identity, cardinality = 1,
+               base_width = 64, reduce_first = 1, dilation = 1,
+               first_dilation = nothing, activation = relu, norm_layer = BatchNorm,
+               drop_block = identity, drop_path = identity,
+               attn_fn = planes -> identity, attn_args::NamedTuple = NamedTuple())
+
+Creates a basic ResNet block.
+
+# Arguments
+
+  - `inplanes`: number of input feature maps
+  - `planes`: number of feature maps for the block
+  - `stride`: the stride of the block
+  - `downsample`: the downsampling function to use
+  - `cardinality`: redundant, kept for compatibility with `bottleneck`.
+  - `base_width`: redundant, kept for compatibility with `bottleneck`.
+  - `reduce_first`: the reduction factor that the input feature maps are reduced by before the first
+    convolution.
+  - `dilation`: the dilation of the second convolution.
+  - `first_dilation`: the dilation of the first convolution.
+  - `activation`: the activation function to use.
+  - `norm_layer`: the normalization layer to use.
+  - `drop_block`: the drop block layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `drop_path`: the drop path layer. This is usually initialised in the `_make_blocks`
+    function and passed in.
+  - `attn_fn`: the attention function to use. See [`squeeze_excite`](#) for an example.
+  - `attn_args`: a NamedTuple that contains none, some or all of the arguments to be passed to the
+    attention function.
+"""
+function basicblock(inplanes, planes; stride = 1, downsample = identity, cardinality = 1,
+                    base_width = 64, reduce_first = 1, dilation = 1,
+                    first_dilation = dilation, activation = relu, norm_layer = BatchNorm,
+                    drop_block = identity, drop_path = identity,
+                    attn_fn = planes -> identity, attn_args::NamedTuple = NamedTuple())
+    expansion = expansion_factor(basicblock)
+    @assert cardinality==1 "`basicblock` only supports cardinality of 1"
+    @assert base_width==64 "`basicblock` does not support changing base width"
+    first_planes = planes ÷ reduce_first
```

Comment:
Oh here they can, but `planes` is required for compatibility with `bottleneck` (and `bottle2neck` in #181)  for calculating the width. So I thought it would be better to just keep a uniform interface

Reply:
Ah that makes sense. Maybe just `reduction_factor` would be a cleaner name?